Ship construction through the power of science!

PowerfulBacon opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->
needs testmerge <3

## About The Pull Request

This PR allows players to research a technology which unlocks devices and machines allowing the construction of custom shuttles.
Shuttles aren't free however, you will need cargo or toxins to get some juicy plasma for your flight.
In the future you will be able to build anchors in space, making the construction of a second station a fun possibility by having a shuttle link between stations (although this would probably require >4h rounds #bringbackoldmrp)
Even if you aren't planning a full-scale station, building a private shuttle for you and your science-bros is extremely satisfying and one of the better parts of the game, so I have made it much more accessible and customisable.
You can even use this update for useful things like putting your tools and gear on the shuttle and flying it around to damaged parts of the stations for repairs as an engineer, as it would be much quicker than flying around and trying to locate a protolathe after you forget something, all round a really nice addition even if you don't plan to ever use this.
<s>You can also fill a shuttle with thrusters and put prisoners or just Ian in for burning alive.</s>

## Why It's Good For The Game

Building shuttles and stuff has always been a fun part of the game but is pretty uncommonly seen. This PR allows (with more difficulty) scientists, and any crewmember to create their own shuttles, which they can use to fly around the station z-level.

This update introduces:
**Custom Shuttle Creator**
A handheld device (reskin of RCD but blue and with a bluespace glow) which, when used, will define an area of up to ~250 (config) tiles to be a custom shuttle. Custom shuttles must be built in space and cannot interact with any part of the station.
Once an area has been selected an airlock must be chosen to act as the docking point and then the shuttle will be created. Admins will recieve a log when a new shuttle is created and only up to 4 can exist at a time.
Shuttle areas require their own APC and power.

To make a shuttle:
 - Build the shell with the airlocks, making sure it's all 1 room and enclosed (and less than <s>200</s> whatever the config is set to tiles).
 - Use the RSD and choose your docking port.
 - Remove some walls and build thrusters + heaters
 - Create your bridge
 - Build whatever else you want including new rooms etc.
 - Fly around

**Custom Shuttle Console**
This is the shuttle controller, the device that allows you to pilot your ships. The custom shuttle console gives you the stats of your shuttle including:
 - Shuttle Mass (Basically how big the shuttle is)
 - Shuttle Driving Force (Combined force of engines)
 - Shuttle Sublight Speed (Force divided by mass (yes I know that's acceleration, but we don't ask questions here))
 - Shuttle Engine Count (How many engines you have)
 - Shuttle Fuel Consumption (The total amount of fuel the shuttle will drain per tile moved)
 - Shuttle Cooldown (How long the shuttle takes to recharge)

To use the console, it must first be paired with a custom shuttle creator, and then the shuttle targetting computer must be used to select a destination for transit.
Once a location has been selected, the computer will calculate the distance, time of travel and fuel cost.

**Shuttle Navigation Computer**
It's a custom navigation computer, much like Whiteship or syndicate nav console, this allows you to select a destination on the station z-level and target the location. Then using the piloting console you can fly to that destination.
To use it must be linked with a custom shuttle creator.

**Plasma Engine Heater**
A special engine heater which can be built with circuit boards from science or engineering. This device requires a piped connection to a pipenet containing plasma. The engine will store plasma for use by a connected plasma engine.
By using upgraded parts more plasma can be stored, or engines can be made more efficient.

**Plasma Engine**
A buildable engine built for custom shuttles, this machine requires a plasma engine heater to be directly connected to function.
The plasma engine will get its fuel from the heater which it will use for travel.
If the required destination needs more fuel per engine than is stored in the plasma engine heater, then the engine will not operation and the speed of the shuttle will be reduced.

**Void Engine**
After alien research has been obtained you can research the void engine. This is the same as the plasma engine, except extremely overpowered and doesn't require fuel.

**Advanced Shuttle Calculations**
Each shuttle has it's own unique properties based on how it has been setup. Small and light shuttles with a lot of thrusters will be extremely fast and throw around everyone not buckled in, which larger shuttles with lower speeds will be much more smooth, but take longer to reach their destination.

## Todo
<s>Route optimisation upgrade - Reduces travel distance by a little bit</s>
<s>Shuttle speed >15 players onboard should be knocked over</s>
<s>Shuttle needs to fly in the direction the console is facing</s>
<s> Shuttle docking port needs to scale correctly so it can actually dock with places (Would allow shuttles to dock with the station and not have them with a fucked up setup)</s>
<s>Shuttle engines should heat up the turf behind them?</s>
<s>Play engine sound on shuttle move</s>
[Scrapped(can do this in another future PR)] <s>Upgraded engine type - Less fuel, faster speed</s>
<s>Void Engine - No fuel consumption ??</s>
[Scrapped(can do this in another future PR)] <s>Bluespace ion thruster - Uses energy, slower, but can travel across z-levels</s>
<s>Deepspace anchor - Allows you to place a marker on another z-level to travel to.</s> (Credit: Cenrus, will just use the existing anchor and make it work around that!)
[Not required, using whiteship ones]<s>Custom sprites for the consoles</s>
<s>Make the docking port direction match the direction of space relative to the door</s>

## Issues
Admin spawned engine heaters don't have their pipe input done correctly until wrenched and rotated, or initialise_directions is varedited to match dir. [This is extremely minor and no matter what I change, this problem still occurs. If you are an abusing admin, after spawning unscrew the engine and rotate 4 times to fix. Sorry 🙁 ]
<s>Possible lagging through spamming calculations (UNTESTED / UNCONFIRMED)</s> Added a cooldown to prevent spamming stat calculations
<s>You might be able to define a shuttle area then go to the station and define that area as your shuttle, allowing you to fly with the station (UNTESTED / UNCONFIRMED)</s> Fixed

## Changelog
:cl:
add: Added in new circuit board for plasma engine
add: Added in new circuit board for plasma engine heater
add: Added in new tool designs for custom shuttle creator
add: Added in new tech branch costing 10000 research points to unlock shuttle creation tools
add: Added the custom shuttle console
add: Added the plasma engine
add: Added the plasma engine heater
add: Added the Custom Shuttle Designator
add: Added the custom shuttle navigation computer
add: Researches to make shuttles more efficient
add: 20% more efficient flight upgrade
add: 40% more efficient flight upgrade
add: 80% more efficient flight upgrade
add: Void thruster - Requires no fuel or heater to operate, produces a very good amount of thrust.
icon: Added in new plasma engine icons
icon: Added in new plasma heater icons
icon: Added icon for shuttle creator
fix: Rather than searching every atom of every turf, now we just store machines that need to be updated in a global list (Credit: Ike709 for idea)
fix: Shuttle can now dock with stationary ports
fix: Shuttles cannot dock on top of plating now
fix: Shuttles can now be made on lavaland
fix: You can no longer target new destinations while in transit
tweak: Shuttle targetter stop working when war is declared
config: Created 2 new config options defining shuttle max size and count of shuttles
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

PowerfulBacon committed:
Removes my local DB settings from the repo (so its in line with beestation not that its sensitive or anything)

PowerfulBacon committed:
This shouldn't be the first push but it adds alot

Adds in the plasma thruster
Adds in the custom camera console
Adds the custom shuttle controller
Adds the Shuttle Area Definer
Adds Research for shuttle area definer

ss13-beebot committed:
Automatic changelog compile [ci skip]

PowerfulBacon committed:
Updated UI with some different text and to look better

PowerfulBacon committed:
Altered plasma engine sprite and added piped engine heater for heating plasma engines

PowerfulBacon committed:
Adds in custom shuttle components, makes a shuttle limit and some other stuff

PowerfulBacon committed:
Adds in the research for shuttle components as well as buildable thrusters

PowerfulBacon committed:
Adds in the research for all the shuttle parts required

PowerfulBacon committed:
Removes debug code

PowerfulBacon committed:
Fixed typo causing crash

PowerfulBacon committed:
Fixed double = issue again

PowerfulBacon committed:
Fixed custom_shuttle referencing engine structure rather than the buildable machine thruster

PowerfulBacon committed:
Fixed missing bracket

PowerfulBacon committed:
Fixed wrong reference

PowerfulBacon committed:
Fixes research panel hiding researches

PowerfulBacon committed:
Made it so the computer boards actually work lmao

PowerfulBacon committed:
Made it so the computer boards can be built in circuit imprinters

PowerfulBacon committed:
Implemented insufficient shuttles engine power, DBZ checks and better failure tooltips

PowerfulBacon committed:
Fixed shuttle time being *10 too long

PowerfulBacon committed:
Player shuttles need power

PowerfulBacon committed:
Adds in engine open hatch sprite

PowerfulBacon committed:
Adds heater_hatch sprite

PowerfulBacon committed:
Updates shuttle items to include hatches

PowerfulBacon committed:
Implmented code to open and close hatch

PowerfulBacon committed:
Added custom sprite, made naming more consistent

PowerfulBacon committed:
Changes a few things to make it work better

PowerfulBacon committed:
Engine updates whether or not it is on, based on whether or not it is connected to an engine heater

PowerfulBacon committed:
Thrusters require a heater to look like they are on

PowerfulBacon committed:
Added admin log for shuttle creation and made the circuit board names consistent with the naming convention

PowerfulBacon committed:
Thruster status affects calculations

PowerfulBacon committed:
Flipped icons, fixed gas pipe direction as wlel as actual fuel consumption lmao

ss13-beebot committed:
Automatic changelog compile [ci skip]

PowerfulBacon committed:
Fixed gas pipe problem, as well as new problem where engines didn't connect

PowerfulBacon committed:
Completed using fuel

PowerfulBacon commented:
Please don't question the fact I forgot to make a branch for this and edited my master, thanks
Also keep an eye out for methods that could be done better please, my main language isn't dreammaker and the source of learning is the rest of the code which obviously is done by hundreds of people and isn't consistent

PowerfulBacon commented:
Also might want to add in something that can detect where shuttle construction is safe, cause I can suspect many players will build in departures only to find the shuttle creation rejected due to there being an area there

PowerfulBacon commented on config/dbconfig.txt:
```diff
@@ -42,4 +42,4 @@ BLOCKING_QUERY_TIMEOUT 5
 
 ## Uncomment to enable verbose BSQL communication logs
-#BSQL_DEBUG
+#BSQL_DEBUG
```
ughhhh why???? do you do this to me

ike709 replied:
Don't worry about it.

PowerfulBacon commented:
Need to add in something to prevent spamming the console buttons of some kind, and checks to make sure were not scanning massive areas (like space) with the digsusting for in a for method

PowerfulBacon commented on code/game/machinery/shuttle/shuttle_heater.dm:
```diff
@@ -0,0 +1,123 @@
+		return
+	//Shitcode omegalul
+	for(var/place in get_area(get_turf(src)))
+		for(var/atom/thing in place)
```
I know there is a better method for this, will have to codedive to find something else

ike709 replied:
I'm pretty half-asleep at the moment so there may be a better way, but you could make a subtype of shuttle machinery for custom shuttles (ex: `/obj/machinery/shuttle/custom/whatever`) and keep track of all of them in a global list, then loop through that list rather than looping through every atom on every turf in the area.

PowerfulBacon replied:
Good idea, thanks

PowerfulBacon commented on code/modules/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,229 @@
+	//Calculate all the data
+	var/list/areas = M.shuttle_areas
+	for(var/shuttleArea in areas)
+		for(var/each in shuttleArea)
```
I know there is a better method for this, will have to codedive to find something else

ike709 replied:
Could just fudge it a bit and go by the number of turfs in the area.

ike709 replied:
But if you want to make it slightly more complex you could consider closed turfs (walls, etc.) as having more mass than open turfs (floors, etc.)

PowerfulBacon replied:
yea currently every single object and turf weighs 100kg

PowerfulBacon replied:
I want machines to weight, so I have it so turfs, machines and structures weigh 100kg (about average for a shuttle's mass to make sense)

ike709 replied:
I'm still not a fan of how expensive looping over literally everything is.

PowerfulBacon replied:
Alright I'll figure something out, have got a spam prevention thing on it

PowerfulBacon replied:
Alright, got a nifty patch up.
Instead of looping through every atom on the turf, we get the amount of atoms on the turf and use that for the weight, only looping through every engine on the turf.

PowerfulBacon commented on code/modules/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,229 @@
+	var/list/areas = M.shuttle_areas
+	//fix this shitcode plox
+	for(var/shuttleArea in areas)
+		for(var/each in shuttleArea)
```
I know there is a better method for this, will have to codedive to find something else

PowerfulBacon replied:
This has to be like this as it is calculating mass too :(

PowerfulBacon replied:
Fixed 🙂

PowerfulBacon commented:
You can select doors that aren't on the shuttle to act as a docking point.
Although this does not break the shuttle and make it massive, it will make the console misaligned and shouldn't be possible.
plz fix <3

PowerfulBacon committed:
Removed comments, removed unused vars

PowerfulBacon commented:
Originally was going to need bluespace crystals for creating shuttles, but that's scrapped

PowerfulBacon committed:
Removed load with bluespace crystals crap

PowerfulBacon commented on code/game/objects/items/shuttle_creator.dm:
```diff
@@ -0,0 +1,253 @@
+	to_chat(user, "<span class='warning'>The [src] bleeps. Select an airlock to create a docking port, or a valid machine to link.</span>")
+	return
+
+/obj/item/shuttle_creator/proc/calculate_bounds(obj/docking_port/mobile/port)
```
This method probably doesn't work at the moment and makes the bounds from a weird point
The size is done, but it needs docking offsets to be set correctly for it to work properly.

ike709 replied:
You can likely figure out the docking offsets by getting the type of each turf going outward from the docking port and iterating an integer until it becomes space or a different area, in both the height direction and width direction.

PowerfulBacon replied:
Thanks, you are awesome

PowerfulBacon commented:
If this needs a vote, would be nice if I could get a response to see if continuing on with this is worth it.
Might do large scale tests at some point too.

Cenrus commented:
We already have "anchors" for shuttles in the form of Bluespace nav beacons, you might not need to code one yourself

ike709 commented on code/game/objects/items/shuttle_creator.dm:
```diff
@@ -0,0 +1,253 @@
+	req_access_txt = "11"
+	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 100, "acid" = 50)
+	resistance_flags = FIRE_PROOF
+	var/static/customShuttles = 0
```
Why is this var static?

PowerfulBacon replied:
To count total number of shuttles created

ike709 replied:
If you want to track all of the shuttles created in the world, it should be a global var. If you just want to track the number made with this specific shuttle creator, it should just be a non-static var.

PowerfulBacon replied:
Alright, will make it global

ike709 commented:
I haven't fully reviewed this (I'm about to go to bed), and there's still some work to be done, but so far this is pretty amazing 👍

PowerfulBacon commented:
> We already have "anchors" for shuttles in the form of Bluespace nav beacons, you might not need to code one 

Ah yes, this should actually actually work once I fix the docking ports bound not aligning up correctly

Although it might be hard powering them in deep space (if they need power idk)

ike709 commented:
> Possible lagging through spamming calculations (UNTESTED / UNCONFIRMED)Test with the handheld tool and spamming maybe

Improving the loops to not check every atom on every turf every time will definitely help with this, but a short cooldown would probably be useful regardless.

ike709 commented:
> Shuttle speed >15 players onboard should be knocked over
> Shuttle needs to fly in the direction the console is facing

These simply requiring editing the correct shuttle vars and/or flags in the correct places.

PowerfulBacon committed:
https://github.com/BeeStation/BeeStation-Hornet/pull/1398#discussion_r396061455

 ike709 1 hour ago • Member

If you want to track all of the shuttles created in the world, it should be a global var. If you just want to track the number made with this specific shuttle creator, it should just be a non-static var.

This does that

PowerfulBacon committed:
Missed a var/static/customShuttles, replaced with global var instead

PowerfulBacon committed:
Instead of searching every atom of every turf in an area, the CheckSetup method now loops through a global list of machines. :tada:

PowerfulBacon committed:
Forgot a /obj

PowerfulBacon committed:
check_setup is now a proc for /obj/machinery/atmospherics/components/unary/shuttle rather than only the heater.

Although this is unlikely to ever be used, it allows for something to be done with other things than heaters that need to be attached or something?

PowerfulBacon committed:
Removed debug line

PowerfulBacon committed:
Shuttle docking bounds are now setup correctly, allowing for docking with stationary ports such as arrivals

PowerfulBacon commented:
movement_force = list("KNOCKDOWN" = 3, "THROW" = 0)

PowerfulBacon committed:
Shuttle can now dock with the auxillery arrivals port

PowerfulBacon committed:
Use OOC filter instead of chat filter for name

PowerfulBacon committed:
Shuttle knocks people over depending on the speed

PowerfulBacon committed:
Fixed clamp not being defined

PowerfulBacon committed:
You can now fly to the aux port

PowerfulBacon committed:
Shuttle flys in the direction of the console

PowerfulBacon committed:
Fixed , instead of ; preventing shuttle from going anywherre

PowerfulBacon committed:
Fixed looping through every single thing to find heaters, and also using atom as a varname

PowerfulBacon committed:
Unused variable removed

PowerfulBacon committed:
(Dirty push to beestation.dme) Adds in new icon and upgrades for shuttle

PowerfulBacon committed:
Revert "(Dirty push to beestation.dme) Adds in new icon and upgrades for shuttle"

This reverts commit 3792f87be0cdac71eb7d8b44076f018bead09c07.

PowerfulBacon committed:
Added shuttle_upgrades.dm with less (only 3 this time) weird line changes

PowerfulBacon commented:
Ignore the massive list, I tried rebranching to fix the 3000 changes to the .dme file but it didn't work.

PowerfulBacon committed:
Added texture for the shuttle disk

PowerfulBacon committed:
Fixed texture filename and also added in the shuttle upgrade disk

PowerfulBacon committed:
Fixed icon again

PowerfulBacon committed:
The problem was with " instead of '

PowerfulBacon committed:
Upgrade disk is no longer consumed by the computer, instead it is reusable

PowerfulBacon committed:
Adds in 2 new upgrades for the shuttle, hyperlane and voidspace

Hyperlane = 0.6x route distance
Voidspace = 0.2x route distance

PowerfulBacon committed:
Dupliate vars removed

PowerfulBacon committed:
Fixed research being bugged and added sprites for different upgrades

PowerfulBacon committed:
Changed text and name of research

PowerfulBacon committed:
Engines now produce heat when used

PowerfulBacon committed:
Implements the void thruster

W..where is the thrust coming f..from?

PowerfulBacon committed:
Void thruster doesn't need to be attatched to a heater to function

PowerfulBacon committed:
Null check for attatched heaters

PowerfulBacon committed:
Fixed duplicate proc def

PowerfulBacon committed:
Shuttles now register built engines, and they also make sound when launching

PowerfulBacon committed:
Play launch sound on launch

PowerfulBacon committed:
Launch is not synced with the sound effect

PowerfulBacon committed:
Slightly lowers engine start time so sound matches better

PowerfulBacon commented:
> 
> 
> > Possible lagging through spamming calculations (UNTESTED / UNCONFIRMED)Test with the handheld tool and spamming maybe
> 
> Improving the loops to not check every atom on every turf every time will definitely help with this, but a short cooldown would probably be useful regardless.

Added in a cooldown on the calculate stats for shuttle (when called by using the UI)

PowerfulBacon committed:
Airlock has to actually be on the shuttle as well as spam prevention systems

PowerfulBacon commented:
All important issues resolved (apart from the admin thing which is only an inconvinience to admins) and everything on the todo list cleared.
Ready for review.

PowerfulBacon committed:
Items don't weight anything

ike709 commented on code/modules/shuttle/shuttle.dm:
```diff
@@ -801,6 +801,10 @@
 			if(!QDELETED(E))
 				engine_list += E
 				. += E.engine_power
+		for(var/obj/machinery/shuttle/engine/E in areaInstance.contents)
```
Notice the for loop directly above this one that also runs.

PowerfulBacon replied:
One is for the structure engines (the current ones which are just scenery) and the other is for my machine thrusters which are buildable by the players

ike709 replied:
Yeah but currently:

1. Both for loops run through the same list with the same type
2. Both are adding different values to the value that gets returned. All of the `engine_power` for all of the engines gets added as well as a `1` for every active thruster.
3. Every engine is getting added to the `engine_list` twice because you're just looping through all of the engines in two separate for loops.

PowerfulBacon replied:
They are of different types one is:
var/obj/structure/shuttle/engine/E the other is 
var/obj/machinery/shuttle/engine/E

ike709 replied:
My bad, I'm blind.

ike709 commented on code/game/machinery/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,60 @@
+			continue
+		if(shuttle_machine == src)
+			continue
+		shuttle_machine.check_setup(FALSE)
```
So every shuttle machine calls every other shuttle machine's check_setup()? Seems like only a single object should be responsible for calling every machine's check_setup().

Also, shouldn't it only run on machines that are on the same shuttle?

PowerfulBacon replied:
The only thing this is used for is checking if an engine has a heater.
Rather then checking everything, I am going to only check the turf that the engine should be on relative to the caller

ike709 commented on code/game/machinery/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,60 @@
+	if(!affectSurrounding)
+		return
+	//Don't update if not on shuttle, to prevent lagging out the server in space
+	if(!istype(get_turf(src), /area/shuttle/custom))
```
Shouldn't this be `get_area(src)`?

PowerfulBacon replied:
Indeed it should be, good spot

ike709 commented on code/game/machinery/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,60 @@
+		shuttle_machine.check_setup(FALSE)
+	//Check the atmospheric devices (The heaters)
+	for(var/obj/machinery/atmospherics/components/unary/shuttle/atmospheric_machine in GLOB.custom_shuttle_machines)
+		if(!atmospheric_machine)
```
This check never fails.

PowerfulBacon replied:
Removed

ike709 commented on code/game/machinery/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,60 @@
+		return
+	//Check the standard machines
+	for(var/obj/machinery/shuttle/shuttle_machine in GLOB.custom_shuttle_machines)
+		if(!shuttle_machine)
```
This check never fails.

PowerfulBacon replied:
Removed

PowerfulBacon committed:
Removed pointless check

PowerfulBacon committed:
Removed another pointless check

ike709 commented on code/game/machinery/shuttle/shuttle_engine.dm:
```diff
@@ -0,0 +1,119 @@
+		return ..()
+	attached_heater = null
+	for(var/atom/thing in heater_turf)
+		if(!istype(thing, /obj/machinery/atmospherics/components/unary/shuttle/heater))
```
You can just use a typed for loop to implicitly run this check (the for loop is already typed anyways).

`for(var/obj/machinery/atmospherics/components/unary/shuttle/heater in heater_turf)`

PowerfulBacon replied:
Changing

ike709 commented on code/game/machinery/shuttle/shuttle_engine.dm:
```diff
@@ -0,0 +1,119 @@
+			continue
+		if(thing.dir != dir)
+			continue
+		var/obj/machinery/atmospherics/components/unary/shuttle/heater/as_heater = thing
```
After changing the loop type, this var becomes unnecessary.

PowerfulBacon replied:
Fixing

ike709 commented on code/game/machinery/shuttle/shuttle_engine.dm:
```diff
@@ -0,0 +1,119 @@
+		return
+	thruster_active = FALSE
+	icon_state = icon_state_off
+	return
```
Trailing returns aren't necessarily a problem, but they are unnecessary (for future reference).

PowerfulBacon replied:
I'll change it, easy fix

PowerfulBacon committed:
Trailing returns removed and for loop fixed

ike709 commented on code/game/machinery/shuttle/shuttle_heater.dm:
```diff
@@ -0,0 +1,136 @@
+	if(!affectSurrounding)
+		return
+	//Don't update if not on shuttle, to prevent lagging out the server in space
+	if(!istype(get_turf(src), /area/shuttle/custom))
```
`get_area(src)`?

PowerfulBacon replied:
Oh missed this, was to prevent people building them in space and lagging server, so less important now but fixing anyway. I think its like this in engine too

ike709 commented on code/game/machinery/shuttle/shuttle_heater.dm:
```diff
@@ -0,0 +1,136 @@
+		return
+		//Check the standard machines
+	for(var/obj/machinery/shuttle/shuttle_machine in GLOB.custom_shuttle_machines)
+		if(!shuttle_machine)
```
Check never fails

PowerfulBacon replied:
Fixing

PowerfulBacon committed:
Pointless check removed

ike709 commented on code/game/machinery/shuttle/shuttle_heater.dm:
```diff
@@ -0,0 +1,136 @@
+	if(!istype(get_turf(src), /area/shuttle/custom))
+		return
+		//Check the standard machines
+	for(var/obj/machinery/shuttle/shuttle_machine in GLOB.custom_shuttle_machines)
```
Another instance in which each machine is calling this on every other machine...?

ike709 commented:
I really don't understand the thought process behind `check_setup()`. What exactly are you trying to accomplish with it? Why do you have it calling itself on other machines? If your goal is to create some kind of linked behavior between machines, such as having an engine be disabled when its heater is, there are better ways to do that.

PowerfulBacon commented:
> 
> 
> I really don't understand the thought process behind `check_setup()`. What exactly are you trying to accomplish with it? Why do you have it calling itself on other machines? If your goal is to create some kind of linked behavior between machines, such as having an engine be disabled when its heater is, there are better ways to do that.

Yea, looking at it originally both engines and heaters were updated and checked, but then it was changed so only the engines check for heaters. Will redo this section

ike709 commented on code/game/machinery/shuttle/shuttle_heater.dm:
```diff
@@ -0,0 +1,136 @@
+		shuttle_machine.check_setup(FALSE)
+	//Check the atmospheric devices (The heaters)
+	for(var/obj/machinery/atmospherics/components/unary/shuttle/atmospheric_machine in GLOB.custom_shuttle_machines)
+		if(!atmospheric_machine)
```
Check never fails.

PowerfulBacon committed:
Pointless check removed

PowerfulBacon commented:
On creation engines need to setup themselves and heaters setup engines around them.
Heaters do not need to be updated themselves, since their setup function does nothing

ike709 commented:
Didn't go through the whole thing, will have another look when you're ready for review again.

PowerfulBacon commented:
Alright, ill fix it up in a bit just finishing a round

PowerfulBacon commented:
Changed it so instead of looping through all the engine heaters and running an update that doesn't do anything causes heaters don't need an update when a heater or engine is edited it only updates the engines.
I am thinking I could just get the turf above, left, right and below and update them only and even if more lines it would be more efficient, so... yea, ill do that instead

PowerfulBacon committed:
fixed get_area being get_turf

PowerfulBacon commented:
Just testing to make sure the same thing is still done, this is SO much better now since we just check 1 turf rather than everything!

PowerfulBacon commented:
Had a lag issue with something causing runtimes I didn't pick up on hence the 2h between comments, just polishing up and submitting soon

PowerfulBacon committed:
Fixed runtime error, fixed engine updating being done in a bad way

PowerfulBacon commented:
@ike709 sorry for the wait, ready for re-review.
Shuttle heater and shuttle engine check_setup() has been changed as well as the updateGasStats() proc in shuttle_heater has had a null check added to prevent adminspawned things from lagging the server.

Gunna go off and sleep now, changes will be delayed :(

PowerfulBacon commented:
nevermind lmao, working find on my end time to debug 🎉
Edit: Fixed, dunno how that worked on my end, must be new byond version

PowerfulBacon committed:
Need travis to rebuild

PowerfulBacon committed:
Removed the parent class, so need to remove calling the parent on return

ike709 commented on code/game/machinery/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,29 @@
+	var/icon_state_open = "burst_plasma_open"
+	var/icon_state_off = "burst_plasma_off"
+
+/obj/machinery/shuttle/Initialize()
```
You also need to make a `Destroy()` subtype that removes it from the list.

ike709 commented on code/game/machinery/shuttle/shuttle_engine.dm:
```diff
@@ -0,0 +1,135 @@
+	var/bluespace_capable = TRUE
+	var/cooldown = 0
+	var/thruster_active = FALSE
+	var/obj/machinery/atmospherics/components/unary/shuttle/heater/attached_heater
```
This should utilize weakrefs. You can find examples in the `code/modules/turbolift` folder or you can ping me on discord and I can explain how they work.

PowerfulBacon replied:
Trying it

ike709 commented on code/game/machinery/shuttle/shuttle_engine.dm:
```diff
@@ -0,0 +1,135 @@
+	air_update_turf()
+
+/obj/machinery/shuttle/engine/attackby(obj/item/I, mob/living/user, params)
+	check_setup()
```
You sure just having this here instead of inside the checks produces the intended behavior?

PowerfulBacon replied:
Yes, if attacked it updates the engine.
Just for the off case of it being something that pushes on attack or something, the sprite should get correctly updated.

ike709 commented on code/game/objects/items/shuttle_upgrades.dm:
```diff
@@ -0,0 +1,52 @@
+/obj/item/shuttle_route_optimisation/attack_obj(obj/O, mob/living/user)
+	. = ..()
+	if(!istype(O, /obj/machinery/computer/custom_shuttle))
+		return
```
Should probably provide feedback to the user that it needs to be used on a shuttle control console. Also it should be clear that this is for custom shuttles, not existing shuttles.

ike709 commented on code/game/objects/items/shuttle_creator.dm:
```diff
@@ -0,0 +1,301 @@
+		return FALSE
+
+	var/datum/map_template/shuttle/new_shuttle = new /datum/map_template/shuttle()
+	var/static/idnum = 0
```
Why is this static?

PowerfulBacon replied:
Fixed, was from the previous static variable

ike709 commented on code/game/objects/items/shuttle_creator.dm:
```diff
@@ -0,0 +1,301 @@
+	return TRUE
+
+/obj/item/shuttle_creator/proc/check_current_area(mob/user)
+	var/static/area_or_turf_fail_types = typecacheof(list(
```
Why is this static?

PowerfulBacon replied:
Because the code is a modified version of /code/__HELPERS/areas.dm /proc/create_area(mob/creator) and it was copied across

ike709 commented on code/modules/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,246 @@
+		return FALSE
+	//Calculate all the data
+	for(var/obj/machinery/shuttle/engine/shuttle_machine in GLOB.custom_shuttle_machines)
+		shuttle_machine.check_setup()
```
You need to check that the shuttle_machine is on the same shuttle as the computer.

PowerfulBacon replied:
Fixed

PowerfulBacon committed:
Fixed shuttle calculating only the weight of objects. Fixed shuttle playing engine sound from other shuttles.

PowerfulBacon committed:
fixed

PowerfulBacon committed:
Changed to give better feedback to the user

PowerfulBacon committed:
Added in destroy subtype

PowerfulBacon committed:
Utilize weakrefs

PowerfulBacon committed:
Fixed weakrefs

PowerfulBacon committed:
Fixed weakrefs

PowerfulBacon committed:
Fixed runtimes

PowerfulBacon committed:
Fixed engine requiring heater even if it doesn't use one

PowerfulBacon committed:
Fixed some sound and runtimes

PowerfulBacon commented:
Almost done pushing and testing the requested changes.

PowerfulBacon committed:
Fixed shuttle weight :( nifty fix lost

PowerfulBacon committed:
Done.

qwertyquerty commented:
I'm about to put this to vote and then we can start testmerges. It would be nice if you could add some GIFs or pictures to your PR desc.

PowerfulBacon commented:
> 
> 
> I'm about to put this to vote and then we can start testmerges. It would be nice if you could add some GIFs or pictures to your PR desc.

Will grab them at the end of the round (almost done). Might take a while to build them since they take some time to build

PowerfulBacon commented:
⭐️

ike709 commented on code/game/machinery/shuttle/shuttle_engine.dm:
```diff
@@ -0,0 +1,138 @@
+	var/obj/machinery/atmospherics/components/unary/shuttle/heater/resolved_heater = attached_heater.resolve()
+	if(panel_open)
+		thruster_active = FALSE
+	else if(resolved_heater.hasFuel(1))
```
```suggestion
	else if(resolved_heater?.hasFuel(1))
```
You need to check that resolved weakrefs are non-null. (`?.` operator like in this usecase is shorthand for `thing && thing.proc_call()` if you weren't aware, it checks if the var is null first)

PowerfulBacon replied:
I see, thanks for the help with this 🎉

ike709 commented on code/modules/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,242 @@
+				continue
+			if(E.attached_heater)
+				var/obj/machinery/atmospherics/components/unary/shuttle/heater/resolvedHeater = E.attached_heater.resolve()
+				if(!resolvedHeater.hasFuel(dist * E.fuel_use) && useFuel)
```
```suggestion
				if(resolvedHeater && !resolvedHeater.hasFuel(dist * E.fuel_use) && useFuel)
```

ike709 commented on code/modules/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,242 @@
+			var/obj/machinery/atmospherics/components/unary/shuttle/heater/resolvedHeater = shuttle_machine.attached_heater.resolve()
+			if(!resolvedHeater.hasFuel(dist * shuttle_machine.fuel_use))
+				continue
+			resolvedHeater.consumeFuel(dist * shuttle_machine.fuel_use)
```
```suggestion
			resolvedHeater?.consumeFuel(dist * shuttle_machine.fuel_use)
```

ike709 commented on code/modules/shuttle/custom_shuttle.dm:
```diff
@@ -0,0 +1,242 @@
+			continue
+		if(shuttle_machine.attached_heater)
+			var/obj/machinery/atmospherics/components/unary/shuttle/heater/resolvedHeater = shuttle_machine.attached_heater.resolve()
+			if(!resolvedHeater.hasFuel(dist * shuttle_machine.fuel_use))
```
```suggestion
			if(resolvedHeater && !resolvedHeater.hasFuel(dist * shuttle_machine.fuel_use))
```

PowerfulBacon committed:
Update code/modules/shuttle/custom_shuttle.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

PowerfulBacon committed:
Update code/modules/shuttle/custom_shuttle.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

PowerfulBacon committed:
Update code/game/machinery/shuttle/shuttle_engine.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

PowerfulBacon commented:
thanks Ike, you are great.
Gotta quickly check this make sure it still works locally just in case

PowerfulBacon committed:
Update code/modules/shuttle/custom_shuttle.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

PowerfulBacon commented:
Alright grabbing screenshots, gimme like 30 mins

PowerfulBacon commented:
![image](https://user-images.githubusercontent.com/26465327/77480836-5c011000-6e1a-11ea-9b9d-e3569c3184d7.png)
![image](https://user-images.githubusercontent.com/26465327/77480895-7cc96580-6e1a-11ea-8516-2bdd56d1a1a4.png)
![image](https://user-images.githubusercontent.com/26465327/77480909-83f07380-6e1a-11ea-92ac-e90b61bbb199.png)
![image](https://user-images.githubusercontent.com/26465327/77481003-aaaeaa00-6e1a-11ea-8c60-438cd1fb90b4.png)

(Bottom one is most likely the reference for what will happen in the round)

PowerfulBacon committed:
Fixed distance upgrade not being in the distance calculator

PowerfulBacon commented:
If the vote passes, don't merge until tomorrow please, need a should be 1 line code change (shuttle prefered flight direction is misalligned) but I can't do it till then. thanks
Also fuel cost too low

PowerfulBacon committed:
Increased fuel cost, made shuttles fly in the right direction

PowerfulBacon commented:
@qwertyquerty Fixed the last remaining things, ready when the vote is done now 🎉

PowerfulBacon committed:
Engine heating power increased a little (Don't put your engines indoors)

Crossedfall commented:
Test merged onto Sage

Crossedfall commented:
Bug reported from test merge: Pirates can teleport into any shuttle being built

PowerfulBacon commented:
Any idea of how to teleport to the shuttle? 

Edit: The ship can jump there or the pirates?
If it's the ship I'll just have to make playing block shuttles like pirates and white ship, as currently you can land on playing and glass with them for some reason

PowerfulBacon commented:
Issue: Shuttles can dock on top of plating, and blobs resulting in my blob core getting deleted.

PowerfulBacon commented:
Issue: Unlinking shuttle midflight results in the shuttle not being able to move

Addendum: Issue is something to do with heaters

PowerfulBacon commented:
<s>Change: Shuttles should show which places they can't go to because of fuel issues</s>
Should already do this

fighterslam commented:
hey can you like
raise the limit to 300
i spent 3 hours makin a kick-ass shuttle but it was 91 over 200
the pain
and the resources

also, you can't launch from lavaland, which is a bit of a pain

Crossedfall commented on code/game/objects/items/shuttle_creator.dm:
```diff
@@ -0,0 +1,299 @@
+#define SHUTTLE_CREATOR_MAX_SIZE 200
+#define CUSTOM_SHUTTLE_LIMIT 4
```
Possible to have these as config options? Would be nice to let server hosts choose how many custom shuttles can be created to scale with their hardware

PowerfulBacon replied:
Will do

PowerfulBacon replied:
(gunna make all the changes tommorow after the test and when I got more time)

PowerfulBacon commented:
Change shuttles to be able to be made on lavaland (add lavaland area to the list of allowed)

PowerfulBacon commented:
Stop custom shuttles docking on plating (do it without changing whiteship if possible)
Or at least windows

PowerfulBacon commented:
Remove the no heater message

PowerfulBacon commented:
![image](https://user-images.githubusercontent.com/26465327/77744835-20309b00-7012-11ea-832a-8d23477583fc.png)
this...
this is bad

PowerfulBacon commented:
Launch timer after pressing button needs to go down by ~5

PowerfulBacon commented:
Shuttles don't fly to evac (thought on this, should they or not)

Crossedfall commented:
Removing test merge for now to allow for fixes and other changes.

TheNeoGamer42 commented:
Yeah with the shuttles there was a WHOLE lotta weirdness, like after a shuttle was there I couldn't put my ship into it and such, it was kinda just weird in general what defined where I can go, but yeah make it so it cant just go onto plating and such

PowerfulBacon committed:
No landing on plating

PowerfulBacon commented:
Turns out the landing on plating is a really easy fix, others shouldn't be hard either.

PowerfulBacon committed:
Lowered shuttle launch timer by 5 so sound lines up better

ss13-beebot committed:
Automatic changelog compile [ci skip]

PowerfulBacon committed:
Shuttles are now a config option

PowerfulBacon committed:
Merge branch 'master' of https://github.com/PowerfulBacon/BeeStation-Hornet-Bacons-Fork

PowerfulBacon committed:
Revert "Merge branch 'master' of https://github.com/PowerfulBacon/BeeStation-Hornet-Bacons-Fork"

This reverts commit ad284164ce4c2466a45638da529e94177ca6c9ef, reversing
changes made to 8da9ada4dbcc1cd46edddd1c694978d456feb41c.

ss13-beebot committed:
Automatic changelog compile [ci skip]

PowerfulBacon committed:
You can now create shuttles on lavaland

PowerfulBacon committed:
Merge branch 'master' of https://github.com/PowerfulBacon/BeeStation-Hornet-Bacons-Fork

PowerfulBacon committed:
Revert "Merge branch 'master' of https://github.com/PowerfulBacon/BeeStation-Hornet-Bacons-Fork"

This reverts commit 64702681bc90d6385a9d1dfe858da3d5f4fbc405, reversing
changes made to 7a01d8c246721578845097a157e20cd82ca8cf9a.

ss13-beebot committed:
Automatic changelog compile [ci skip]

PowerfulBacon commented:
Why does beebot keep changing shit in the config file?

PowerfulBacon committed:
Revert "Automatic changelog compile [ci skip]"

This reverts commit 658c959ca3f5b6c760fe193e7ed68bb3d297ddc2.

ss13-beebot committed:
Automatic changelog compile [ci skip]

PowerfulBacon commented:
> 
> 
> hey can you like
> raise the limit to 300
> i spent 3 hours makin a kick-ass shuttle but it was 91 over 200
> the pain
> and the resources
> 
> also, you can't launch from lavaland, which is a bit of a pain

Config option now, so up to the server host

PowerfulBacon committed:
Custom shuttles use correct circuitboard

PowerfulBacon committed:
Merge branch 'master' of https://github.com/PowerfulBacon/BeeStation-Hornet-Bacons-Fork

PowerfulBacon commented:
> 
> 
> Issue: Unlinking shuttle midflight results in the shuttle not being able to move
> 
> Addendum: Issue is something to do with heaters

This might have been an issue with them losing the thing and vareditting shuttleId , but the shuttle heater warning comes up

Bug is to do with moving the shuttle location while in transit

PowerfulBacon committed:
Fix: Shuttles can no longer target destinations while in transit softlocking the shuttle into being stuck

PowerfulBacon committed:
Custom shuttles don't work when war is declared

PowerfulBacon commented:
@Crossedfall Fixed all of the bugs and issues that were reported that I could find, although I cannot find or replicate the cause of the pirate teleporting bug, It Is possible for the pirate ship to jump to a shuttle being build (since it is just plating) but actually teleporting the people I haven't figured out

Also beebot is forcing changeling commits that ok?

PowerfulBacon committed:
Debug message removed

PowerfulBacon committed:
Reverted lavaland shuttles because they break lavaland

MNarath1 commented:
So is this now going to stay after the bug fixes for golden and sage?

TheNeoGamer42 commented:
2nd testmerge when plz giv soon

qwertyquerty reviewed (approved):
@crossedfall I have reviewed all changes and pronounce this good for the next testmerge

qwertyquerty commented:
Testmerged onto sage

Crossedfall commented:
Gotcha, might have been a fluke? Not entirely sure myself. 

As for beebot, that's a bit strange. I think it has to do with how you've been merging and reverting the master into your branch. You should be rebasing ontop of the master instead, especially for something this chunky. Before this is merged, you'll need to remove the CL commits made by beebot. I can help with that if needed

PowerfulBacon commented:
> 
> 
> Gotcha, might have been a fluke? Not entirely sure myself.
> 
> As for beebot, that's a bit strange. I think it has to do with how you've been merging and reverting the master into your branch. You should be rebasing ontop of the master instead, especially for something this chunky. Before this is merged, you'll need to remove the CL commits made by beebot. I can help with that if needed

Thanks, I'll give it a shot but ill ping on you discord if I can't get it fixed

PowerfulBacon committed:
Merge remote-tracking branch 'upstream/master'

PowerfulBacon commented:
> 
> 
> Gotcha, might have been a fluke? Not entirely sure myself.
> 
> As for beebot, that's a bit strange. I think it has to do with how you've been merging and reverting the master into your branch. You should be rebasing ontop of the master instead, especially for something this chunky. Before this is merged, you'll need to remove the CL commits made by beebot. I can help with that if needed

Fixed the changelog changes, thanks for the help

PowerfulBacon committed:
thing to make travis recompile

PowerfulBacon committed:
Revert "Merge remote-tracking branch 'upstream/master'"

This reverts commit fc1bb8e7b4ccc4463fac21bb18474bc29138f292, reversing
changes made to ad7bb856e1f310890024c94c36af3cd7b5bde83d.

PowerfulBacon committed:
Revert "Revert "Merge remote-tracking branch 'upstream/master'""

This reverts commit fde6a6971beeed67ccb3a72e72793987daa6646f.

PowerfulBacon committed:
Merge remote-tracking branch 'upstream/master'

MarkSuckerberg commented on code/game/objects/items/shuttle_creator.dm:
```diff
@@ -0,0 +1,299 @@
+		to_chat(user, "<span class='warning'>The provided ship name is too long, blares the [src]</span>")
+		return FALSE
+	if(OOC_FILTER_CHECK(str))
+		to_chat(user, "<span class='warning'>Nanotrasen prohibited words are in use in this shuttle name, blars the [src] in a slightly offended tone.</span>")
```
typo, forgot an e at the end of blars

MarkSuckerberg commented on code/game/objects/items/shuttle_creator.dm:
```diff
@@ -0,0 +1,299 @@
+		to_chat(user, "<span class='warning'>Shuttles must be created in an airtight space, ensure that the shuttle is airtight, including corners.</span>")
+		return
+	if(turfs.len > SHUTTLE_CREATOR_MAX_SIZE)
+		to_chat(user, "<span class='warning'>The [src]'s internal cooling system wizzes violently and a message appears on the screen, \"Caution, this device can only handle the creation of shuttles up to [SHUTTLE_CREATOR_MAX_SIZE]. Please reduce your shuttle by [turfs.len-SHUTTLE_CREATOR_MAX_SIZE]. Sorry for the inconvinience\"</span>")
```
You should have it say turfs after the amount of turfs, so that it says "can only create shuttles up to [x] turfs in size," instead of "can only create shuttles up to [x]."

MarkSuckerberg commented on code/game/objects/items/shuttle_upgrades.dm:
```diff
@@ -0,0 +1,55 @@
+	throw_range = 5
+	density = FALSE
+	anchored = FALSE
+	item_flags = NOBLUDGEON
```
don't need to re-define all this for this if it's parent has exactly the same stats

MarkSuckerberg reviewed (commented):
Just some things I noticed at first glance, mostly grammatical because ~~I'm an asshole~~ grammar is inportant

Isy232 commented:
Perhaps bring down the point cost of shuttle tech, 15000 is a lot of points just for shuttles and it forces shuttles to only be possible to make later into the round when you have lots of points spare.

PowerfulBacon committed:
Removes repeated defines

ss13-beebot committed:
Automatic changelog compile [ci skip]

PowerfulBacon committed:
Fixes spelling and grammar

PowerfulBacon committed:
Reduced shuttle research cost to 10000, to make them still later game but at the same time not halting research for too long.

PowerfulBacon committed:
Merge branch 'master' of https://github.com/PowerfulBacon/BeeStation-Hornet-Bacons-Fork

PowerfulBacon committed:
Merge remote-tracking branch 'upstream/master'

qwertyquerty commented:
This ready for another testmerge?

PowerfulBacon commented:
> This ready for another testmerge?

Should be, almost the same as before but with some grammatical fixes

ike709 commented:
Please create a guide on the wiki for building constructable shuttles, then this PR should be good for merge if no new issues have arisen,

PowerfulBacon commented:
> Please create a guide on the wiki for building constructable shuttles, then this PR should be good for merge if no new issues have arisen,

Will do

PowerfulBacon commented:
Slowly making progress on the wiki page, give me a while and it'll be done <s>eventually</s>

qwertyquerty commented:
@PowerfulBacon how are those wiki changes coming along

PowerfulBacon commented:
> 
> 
> @PowerfulBacon how are those wiki changes coming along

Good, been working on this so far
https://wiki.beestation13.com/view/User:PowerfulBacon/Sandbox#Shuttle_Machinery

PowerfulBacon commented:
https://wiki.beestation13.com/view/Guide_to_shuttle_construction

ike709 commented:
Conflicts.

PowerfulBacon committed:
Merge branch 'master' into master

ike709 commented on config/Sage/config.txt:
```diff
@@ -497,6 +497,8 @@ MAX_SLIMES 60
 MAX_SHUTTLE_COUNT 6
 MAX_SHUTTLE_SIZE 200
 
+=======
```
```suggestion
```

ike709 commented on config/config.txt:
```diff
@@ -496,6 +496,8 @@ MAX_SLIMES 60
 MAX_SHUTTLE_COUNT 6
 MAX_SHUTTLE_SIZE 200
 
+=======
```
```suggestion
```

PowerfulBacon committed:
Update config/Sage/config.txt

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

PowerfulBacon commented:
Just going to run it locally, make sure it works

PowerfulBacon committed:
Update config/config.txt

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

ike709 commented on config/Sage/config.txt:
```diff
@@ -488,6 +488,16 @@ GRANT_METACURRENCY
 ## Comment this out to ignore global bans
 RESPECT_GLOBAL_BANS
 
+## Mob spam prevention. The number of each mobtype that can be alive at once. Stops people from crashing the server with chickens/monkeycubes/slimes. Altering these values is recommended based on server hardware.
```
Remove this whole section you re-added.

ike709 commented on config/config.txt:
```diff
@@ -487,6 +487,16 @@ GRANT_METACURRENCY
 ## Comment this out to ignore global bans
 RESPECT_GLOBAL_BANS
 
+## Mob spam prevention. The number of each mobtype that can be alive at once. Stops people from crashing the server with chickens/monkeycubes/slimes. Altering these values is recommended based on server hardware.
```
Remove this whole section you re-added.

PowerfulBacon committed:
Removed unused part of config that I readded because I am a pepaga

Crossedfall commented on config/Sage/config.txt:
```diff
@@ -488,6 +488,11 @@ GRANT_METACURRENCY
 
+## Custom shuttle spam prevention. Changine these numbers allows you to change the maxsize and amount of custom shuttles.
+MAX_SHUTTLE_COUNT 6
+MAX_SHUTTLE_SIZE 200
```
```suggestion
MAX_SHUTTLE_SIZE 300
```

Crossedfall commented on config/config.txt:
```diff
@@ -487,6 +487,11 @@ GRANT_METACURRENCY
 
+## Custom shuttle spam prevention. Changine these numbers allows you to change the maxsize and amount of custom shuttles.
+MAX_SHUTTLE_COUNT 6
+MAX_SHUTTLE_SIZE 200
```
```suggestion
MAX_SHUTTLE_SIZE 300
```

Crossedfall reviewed (approved):
Good to go, but I have these set to 300 on the server since testing seemed to be fine.

PowerfulBacon committed:
Update config/config.txt

Co-Authored-By: Crossedfall <ping@crossedfall.io>

PowerfulBacon committed:
Update config/Sage/config.txt

Co-Authored-By: Crossedfall <ping@crossedfall.io>
